sdk/rust/oak/src/handle.rs

```diff
@@ -0,0 +1,224 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Utilities for visiting, extract and injecting handles.
+//!
+//! Applications will usually not interact with the types in this module directly, as the
+//! `HandleVisit` trait is automatically derived for all protobuf types compiled with `oak_utils`,
+//! and extracting and injecting handles is taken care of by
+//! [`oak::io::Receiver`](../io/struct.Receiver.html) and
+//! [`oak::io::Sender`](../io/struct.Sender.html).
+include!(concat!(env!("OUT_DIR"), "/oak.handle.rs"));
+
+use crate::OakError;
+
+/// Raw handle identifier.
+///
+/// This is only meant to be used with the [HandleVisit](trait.HandleVisit.html) trait, for
+/// anything else [Handle](../struct.Handle.html) should be preferred.
+// Note to maintainers: users may not add oak_abi as an explicit dependency, so we re-export this
+// type here for the custom derive to use.
+pub type Handle = oak_abi::Handle;
+
+/// Visit all handles present in a type.
+///
+/// The most common types that contains handles are
+/// [`oak::io::Receiver`](../io/struct.Receiver.html) and
+/// [`oak::io::Sender`](../io/struct.Sender.html).
+///
+/// This trait can be automatically derived:
+/// ```
+/// use oak::handle::HandleVisit;
+///
+/// #[derive(HandleVisit)]
+/// struct Thing {
+///     receiver: oak::io::Receiver<String>,
+///     sender: oak::io::Sender<String>,
+/// }
+/// ```
+///
+/// Alternatively, you can implement the trait manually. This is required if you want the trait to
+/// visit handles that are directly contained in a type.
+///
+/// ```
+/// use oak::handle::{Handle, HandleVisit};
+///
+/// struct Thing {
+///     handle: Handle,
+/// }
+///
+/// impl HandleVisit for Thing {
+///     fn visit<F: FnMut(&mut Handle)>(&mut self, mut visitor: F) -> F {
+///         visitor(&mut self.handle);
+///         visitor
+///     }
+/// }
+/// ```
+pub trait HandleVisit {
+    /// Invokes the provided closure on every handle contained in `self`.
```

Comment:
Comment on why it's `&mut`?

Reply:
I've added a comment for this to the `visit` method.